### PR TITLE
[Metrics] [Docs] Add section about device placement

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4909
Adds a section to metrics docs about device placement

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes [if needed]?
- [ ] Did you write any new necessary tests [no need for typos, docs]? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone are aligned!**
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+    target = torch.tensor([1, 1, 0, 0], device=torch.device("cuda", 0))
+    preds = torch.tensor([0, 1, 0, 0], device=torch.device("cuda", 0))
+
+    # Metric states are always initialized on cpu, and needs to be moved to
+    # the correct device
+    confmat = Accuracy(num_classes=2).to(torch.device("cuda", 0))
+    out = confmat(preds, target)
+    print(out.device) # cuda:0
+
+However, when **proper defined** inside an ``LightningModule``, lightning will
```
Comment (Shreeyak) [inline docs/source/metrics.rst:162]:
```suggestion
However, when **properly defined** inside an ``LightningModule``, lightning will
```

File: docs/source/metrics.rst [hunk 2]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+automatically move the metrics to the same device as the data. By **proper defined**
```
Comment (Shreeyak) [inline docs/source/metrics.rst:163]:
```suggestion
automatically move the metrics to the same device as the data. By **properly defined**
```

File: docs/source/metrics.rst [hunk 3]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+it is meant that the metric is correctly identified as a child module of the model 
+(check ``.children()`` attribute of the model). Therefore, metrics cannot be placed 
+in native python ``list`` and ``dict``, as they will not be correctly identified
+as a child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of
```
Comment (Shreeyak) [inline docs/source/metrics.rst:167]:
```suggestion
as child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of
```

**Review by Shreeyak — COMMENTED** [review]
Beautiful. Thanks for adding these details. I narrowly avoided using a native dict through luck.

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+    target = torch.tensor([1, 1, 0, 0], device=torch.device("cuda", 0))
+    preds = torch.tensor([0, 1, 0, 0], device=torch.device("cuda", 0))
+
+    # Metric states are always initialized on cpu, and needs to be moved to
+    # the correct device
+    confmat = Accuracy(num_classes=2).to(torch.device("cuda", 0))
+    out = confmat(preds, target)
+    print(out.device) # cuda:0
+
+However, when **properly defined** inside an ``LightningModule``, lightning will
```
Comment (s-rog) [inline docs/source/metrics.rst:162]:
```suggestion
However, when **properly defined** inside a ``LightningModule``, lightning will
```

File: docs/source/metrics.rst [hunk 5]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+automatically move the metrics to the same device as the data. By **properly defined**
```
Comment (s-rog) [inline docs/source/metrics.rst:163]:
```suggestion
automatically move the metrics to the same device as the data. Being **properly defined**
```

File: docs/source/metrics.rst [hunk 6]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+it is meant that the metric is correctly identified as a child module of the model
```
Comment (s-rog) [inline docs/source/metrics.rst:164]:
```suggestion
means that the metric is correctly identified as a child module of the model 
```

**Review by s-rog — APPROVED** [review]

File: docs/source/metrics.rst [hunk 7]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+
+However, when **properly defined** inside a ``LightningModule``, lightning will
+automatically move the metrics to the same device as the data. Being **properly defined**
+means that the metric is correctly identified as a child module of the model 
+(check ``.children()`` attribute of the model). Therefore, metrics cannot be placed 
+in native python ``list`` and ``dict``, as they will not be correctly identified
+as child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of
+``dict`` use ``torch.nn.ModuleDict``.
+
+.. code-block::
```
Comment (Borda) [inline docs/source/metrics.rst:170]: can this be test too?
```suggestion
.. testcode::
```

**Review by Borda — APPROVED** [review]

File: docs/source/metrics.rst [hunk 8]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+    target = torch.tensor([1, 1, 0, 0], device=torch.device("cuda", 0))
+    preds = torch.tensor([0, 1, 0, 0], device=torch.device("cuda", 0))
+
+    # Metric states are always initialized on cpu, and needs to be moved to
+    # the correct device
+    confmat = Accuracy(num_classes=2).to(torch.device("cuda", 0))
+    out = confmat(preds, target)
+    print(out.device) # cuda:0
+
+However, when **properly defined** inside a ``LightningModule``, lightning will
```
Comment (awaelchli) [inline docs/source/metrics.rst:162]:
```suggestion
However, when **properly defined** inside a ``LightningModule``, Lightning will
```
I think generally the Lightning brand is capitalized.

File: docs/source/metrics.rst [hunk 9]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+automatically move the metrics to the same device as the data. Being **properly defined**
+means that the metric is correctly identified as a child module of the model 
+(check ``.children()`` attribute of the model). Therefore, metrics cannot be placed 
+in native python ``list`` and ``dict``, as they will not be correctly identified
+as child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of
```
Comment (awaelchli) [inline docs/source/metrics.rst:167]:
```suggestion
as child modules. Instead of ``list`` use :class:`torch.nn.ModuleList` and instead of
```
btw we can directly link torch docs. Hope I didn't make a typo

File: docs/source/metrics.rst [hunk 10]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+``dict`` use ``torch.nn.ModuleDict``.
```
Comment (awaelchli) [inline docs/source/metrics.rst:168]:
```suggestion
``dict`` use :class:`torch.nn.ModuleDict`.
```

**Review by awaelchli — APPROVED** [review]

File: docs/source/metrics.rst [hunk 11]
```diff
@@ -137,6 +137,56 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
     To change this, after initializing the metric, the method ``.persistent(mode)`` can
     be used to enable (``mode=True``) or disable (``mode=False``) this behaviour.
 
+*******************
+Metrics and devices
+*******************
+
+Metrics are simple subclasses of ``torch.nn.Module`` and their metric states behave
```
Comment (rohitgr7) [inline docs/source/metrics.rst:144]:
```suggestion
Metrics are simple subclasses of :class:`~torch.nn.Module` and their metric states behave
```

**Review by rohitgr7 — APPROVED** [review]

File: docs/source/metrics.rst (hunk 9 above)
Comment (rohitgr7) [reply docs/source/metrics.rst:167]: just missed `~`

File: docs/source/metrics.rst [hunk 12]
```diff
@@ -137,6 +137,57 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
 
+*******************
+Metrics and devices
+*******************
+
+Metrics are simple subclasses of :class:`~torch.nn.Module` and their metric states behave
+similar to buffers and parameters of modules. This means that metrics states should
+be moved to the same device as the input of the metric:
+
+.. code-block:: python
```
Comment (rohitgr7) [inline docs/source/metrics.rst:148]:
```suggestion
.. testcode::
```
I hope it's correct.

File: docs/source/metrics.rst [hunk 13]
```diff
@@ -137,6 +137,57 @@ This metrics API is independent of PyTorch Lightning. Metrics can directly be us
...
+    # the correct device
+    confmat = Accuracy(num_classes=2).to(torch.device("cuda", 0))
+    out = confmat(preds, target)
+    print(out.device) # cuda:0
+
+However, when **properly defined** inside a :class:`~pytorch_lightning.core.lightning.LightningModule`
+, Lightning will
+automatically move the metrics to the same device as the data. Being **properly defined**
+means that the metric is correctly identified as a child module of the model 
+(check ``.children()`` attribute of the model). Therefore, metrics cannot be placed
```
Comment (rohitgr7) [inline docs/source/metrics.rst:166]:
```suggestion
, Lightning will automatically move the metrics to the same device as the data. Being
**properly defined** means that the metric is correctly identified as a child module of the
model (check ``.children()`` attribute of the model). Therefore, metrics cannot be placed
```

File: docs/source/metrics.rst (hunk 9 above)
Comment (awaelchli) [reply docs/source/metrics.rst]: it's optional, it depends if you want to show short or long name :)
